crates/cdk/src/wallet/proofs.rs

```diff
@@ -166,112 +167,393 @@ impl Wallet {
+                &active_proofs,
+                amount.checked_sub(selected_amount).unwrap_or(Amount::ZERO),
+                keyset_fees,
+            ) {
+                selected_proofs.extend(proofs);
+                return Ok(selected_proofs);
+            }
+        }
 
-            let proof_amount = selected_proof.amount;
-
-            selected_proofs.push(selected_proof);
+        sort_proofs(&mut active_proofs, opts.method, amount);
 
-            let fees = match include_fees {
-                true => self.get_proofs_fee(&selected_proofs).await?,
-                false => Amount::ZERO,
+        for active_proof in active_proofs {
+            selected_proofs.push(active_proof);
+            let selected_total = Amount::try_sum(selected_proofs.iter().map(|p| p.amount))?;
+            let fees = if opts.include_fees {
+                self.get_proofs_fee(&selected_proofs).await?
+            } else {
+                Amount::ZERO
             };
 
-            if proof_amount >= remaining_amount + fees {
-                remaining_amount = Amount::ZERO;
-                break;
+            if selected_total >= amount + fees {
+                return Ok(selected_proofs);
             }
+        }
+
+        Err(Error::InsufficientFunds)
+    }
+}
 
-            remaining_amount = amount.checked_add(fees).ok_or(Error::AmountOverflow)?
-                - Amount::try_sum(selected_proofs.iter().map(|p| p.amount))?;
-            (proofs_larger, proofs_smaller) = proofs_smaller
-                .into_iter()
-                .skip(1)
-                .partition(|p| p.amount > remaining_amount);
+fn sort_proofs(proofs: &mut Proofs, method: ProofSelectionMethod, amount: Amount) {
+    match method {
+        // Least fallback to largest
+        ProofSelectionMethod::Largest | ProofSelectionMethod::Least => {
+            proofs.sort_by(|a: &Proof, b: &Proof| b.cmp(a))
         }
+        ProofSelectionMethod::Closest => proofs.sort_by(|a: &Proof, b: &Proof| {
+            let a_diff = if a.amount > amount {
+                a.amount - amount
+            } else {
+                amount - a.amount
+            };
+            let b_diff = if b.amount > amount {
+                b.amount - amount
+            } else {
+                amount - b.amount
+            };
+            a_diff.cmp(&b_diff)
+        }),
+        ProofSelectionMethod::Smallest => proofs.sort(),
+    }
+}
 
-        if remaining_amount > Amount::ZERO {
-            if let Some(next_bigger) = next_bigger_proof {
-                return Ok(vec![next_bigger.clone()]);
+fn select_least_proofs_over_amount(
+    proofs: &Proofs,
+    amount: Amount,
+    fees: HashMap<Id, u64>,
+) -> Option<Vec<Proof>> {
+    tracing::trace!(
+        "Selecting LEAST proofs over amount {} with fees {:?}",
+        amount,
+        fees
+    );
+    let max_sum = Amount::try_sum(proofs.iter().map(|p| p.amount)).ok()? + 1.into();
```

Comment:
Since `1` is added outside of `try_sum`, it could overflow.

IMO it should be put in `try_sum`:
```rust
let mut addends: Vec<Amount> = proofs.iter().map(|p| p.amount).collect();
addends.push(1.into());
let max_sum = Amount::try_sum(addends).ok()?;
```

Reply:
It will panic, not overflow since the amount type internally does a checked add with an expect. https://github.com/cashubtc/cdk/blob/260f262c395056ae243ccc8cfa0e9de793c5fd20/crates/cdk/src/amount.rs#L169-L175. But could use a checked add if we want to be explicit about it.